### PR TITLE
For telemetry Add IMDSv2 along with IMDSv1

### PR DESCRIPTION
**Pull request by jeet4320**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
For telemetry, we use IMDSv1 now which won't work if instance is configured to have http_tokens as required for which IMDSv2 is needed. So adding IMDSv2 as well to retrieve instance metadata if we cant get it from IMDSv1. Note that IMDSv2 won't work if instance hop limit is 1, so have added tests for it as well.

Moreover, to avoid throttling at IMDSv2 level, using TOKEN once per session instead of generating TOKEN for each request.
More details as given below:

For IMDSv1, there is no practical limit on the number of concurrent tokens, each representing its own session. IMDSv2 is, however, still constrained by normal instance metadata service connection and throttling limits. For more information, see [Query throttling](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instancedata-throttling) (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instancedata-throttling). [[reference](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html) (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html)]

We throttle queries to the instance metadata service on a per-instance basis, and we place limits on the number of simultaneous connections from an instance to the instance metadata service.  If you're using the instance metadata service to retrieve AWS security credentials, avoid querying for credentials during every transaction or concurrently from a high number of threads or processes, as this might lead to throttling. Instead, we recommend that you cache the credentials until they start approaching their expiry time. If you are throttled while accessing the instance metadata service, retry your query with an exponential backoff strategy. [[reference](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instancedata-throttling) (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instancedata-throttling)]

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: mxnet/buildspec.yml [hunk 1]
```diff
@@ -54,113 +54,3 @@ images:
     target: ec2
     context:
       <<: *TRAINING_CONTEXT
-  BuildEC2MXNetGPUTrainPy3DockerImage:
```
Comment (jeet4320) [inline mxnet/buildspec.yml:57]: to be reverted, commented to test one image

File: dlc_developer_config.toml [hunk 2]
```diff
@@ -23,9 +23,9 @@ benchmark_mode = false
 [build]
 # Frameworks for which you want to disable both builds and tests
 # available frameworks - ["autogluon", "huggingface_tensorflow", "huggingface_pytorch", "huggingface_tensorflow_trcomp", "huggingface_pytorch_trcomp", "tensorflow", "mxnet", "pytorch"]
-skip_frameworks = []
+skip_frameworks = ["autogluon", "huggingface_tensorflow" , "huggingface_pytorch", "huggingface_tensorflow_trcomp", "huggingface_pytorch_trcomp", "tensorflow", "pytorch"]
```
Comment (jeet4320) [inline dlc_developer_config.toml:26]: this file is to be reverted

File: src/deep_learning_container.py [hunk 3]
```diff
@@ -24,6 +24,64 @@
...
+        response_text = response.text
+
+    return response_text
+
+def initialize_imdsv2_token():
+    response = None
+    headers = {"X-aws-ec2-metadata-token-ttl-seconds": "600"}
+    url = "http://169.254.169.254/latest/api/token"
+    timeout = TIMEOUT_SECS
+    global TOKEN
```
Comment (arjkesh) [inline src/deep_learning_container.py:72]: nit: use of globals here will likely get confusing and potentially lead to race conditions. Why not return the token here, and pass it in as an argument to request_helper_imds?

File: src/deep_learning_container.py [hunk 4]
```diff
@@ -24,6 +24,64 @@
...
+    if TOKEN is None:
```
Comment (arjkesh) [inline src/deep_learning_container.py:73]: in what cases will the token be set before this runs?

File: src/deep_learning_container.py [hunk 5]
```diff
@@ -24,6 +24,64 @@
 
 TIMEOUT_SECS = 5
 
+TOKEN = None
+
+def requests_helper(url, headers = None, timeout = 0.1):
```
Comment (arjkesh) [inline src/deep_learning_container.py:29]: can you provide descriptions on the functions added - some of the names are pretty similar so it would be helpful to have high level descriptions

File: src/deep_learning_container.py [hunk 6]
```diff
@@ -44,21 +102,25 @@ def _retrieve_instance_id():
...
-    # We can't use IMDSv2 here which needs token, as adding it mandates docker container to add additional parameter "--network host", else it hangs.
-    url = "http://169.254.169.254/latest/meta-data/instance-id"
-    response = requests_helper(url, timeout=0.1)
+    instance_url = "http://169.254.169.254/latest/meta-data/instance-id"
+    
+    instance_id = requests_helper_imds(instance_url)
 
-    if response is not None and not (400 <= response.status_code < 600):
-        instance_id = _validate_instance_id(response.text)
+    if instance_id is None:
```
Comment (arjkesh) [inline src/deep_learning_container.py:109]: do you need to explicitly check for "None" or is empty string or 0 allowed? Those cases will be missed in the two if statements.

Additionally, why are these two separate if statements instead of if/else?

File: src/deep_learning_container.py [hunk 7]
```diff
@@ -77,12 +139,17 @@ def _retrieve_instance_region():
...
-    url = "http://169.254.169.254/latest/dynamic/instance-identity/document"
-    response = requests_helper(url, timeout=0.1)
 
-    if response is not None and not (400 <= response.status_code < 600):
-        response_json = json.loads(response.text)
+    region_url = "http://169.254.169.254/latest/dynamic/instance-identity/document"
+
+    response_text = requests_helper_imds(region_url)
+
+    if response_text is None:
```
Comment (arjkesh) [inline src/deep_learning_container.py:147]: Same question as above regarding if statements

File: test/dlc_tests/sanity/test_telemetry.py [hunk 8]
```diff
@@ -193,7 +203,100 @@ def _run_tag_success(image_uri, ec2_client, ec2_instance, ec2_connection):
 
     preexisting_ec2_instance_tags = ec2_utils.get_ec2_instance_tags(ec2_instance_id, ec2_client=ec2_client)
     if expected_tag_key in preexisting_ec2_instance_tags:
-        ec2_client.remove_tags(Resources=[ec2_instance_id], Tags=[{"Key": expected_tag_key}])
+        ec2_client.delete_tags(Resources=[ec2_instance_id], Tags=[{"Key": expected_tag_key}])
+
+    if "tensorflow" in framework and job_type == "inference":
```
Comment (arjkesh) [inline test/dlc_tests/sanity/test_telemetry.py:208]: Did this piece come from anywhere or did we never properly test tf inference?

File: test/test_utils/ec2.py [hunk 9]
```diff
@@ -459,6 +460,38 @@ def get_ec2_instance_tags(instance_id, region=DEFAULT_REGION, ec2_client=None):
     return {tag["Key"]: tag["Value"] for tag in response.get("Tags")}
 
 
+def enforce_IMDSv2(instance_id, region=DEFAULT_REGION, ec2_client=None, hop_limit = 1):
```
Comment (arjkesh) [inline test/test_utils/ec2.py:463]: please add function description here

File: test/dlc_tests/sanity/test_telemetry.py (hunk 8 above)
Comment (jeet4320) [reply test/dlc_tests/sanity/test_telemetry.py:208]: Its a copy paste code from other func in same file https://github.com/aws/deep-learning-containers/pull/2239/files/32a48d8273dcdecc286f38326003c1fafb075a15#diff-a80f3063ba4837df91fd3c0137d220e46a795f7805f76ba6ac8806d428312f44R158

Comment (jeet4320) [reply test/dlc_tests/sanity/test_telemetry.py:208]: let me try to reuse it by defining it in func

**Comment by arjkesh** [comment]
Thank you for making the changes - looks like there are some merge conflicts to address but changes LGTM

File: test/dlc_tests/sanity/test_telemetry.py [hunk 10]
```diff
@@ -142,38 +154,26 @@ def _run_instance_role_disabled(image_uri, ec2_client, ec2_instance, ec2_connect
...
-        assert output.ok, f"'import {framework_to_import}' fails when credentials not configured"
+    import_framework(image_uri, container_name, docker_cmd, framework, job_type, ec2_connection)
 
     ec2_instance_tags = ec2_utils.get_ec2_instance_tags(ec2_instance_id, ec2_client=ec2_client)
     assert expected_tag_key not in ec2_instance_tags, (
         f"{expected_tag_key} was applied as an instance tag."
         "EC2 create_tags went through even though it should not have"
     )
+    # Enable access to EC2 instance metadata, so other tests can be run on same EC2 instance
+    ec2_connection.run(f"sudo route del -host 169.254.169.254 reject")
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_telemetry.py:170]: Good call.

**Review by saimidu — DISMISSED** [review]

**Review by saimidu — APPROVED** [review]

File: test/test_utils/ec2.py [hunk 11]
```diff
@@ -459,6 +460,47 @@ def get_ec2_instance_tags(instance_id, region=DEFAULT_REGION, ec2_client=None):
...
+        HttpPutResponseHopLimit=hop_limit,
+        HttpEndpoint='enabled',
+    )
+
+    if not response:
+        raise Exception("Unable to enforce IMDSv2. No response received.")
+
+    timeout = 3
+    state = None
+    if response["InstanceId"]:
```
Comment (ohadkatz) [inline test/test_utils/ec2.py:486]: May cause unexpected KeyError if InstanceId not in response. Might be beneficial to use response.get("InstanceId") to avoid unexpected errors

**Review by ohadkatz — APPROVED** [review]
